### ISSUE TITLE
Investigate cause of 0.37.1

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#7322]
Certbot 0.37.1 was released last week reverting the change we made to disable TLS session tickets in Apache as it caused TLS errors on some systems. More details about the failures can be found at https://community.letsencrypt.org/t/ssl-error-after-cert-renew/99430 and our public Mattermost channel.

I think we should investigate the cause of these errors. As of right now, no one on the Certbot team has been able to reproduce them. The reasons I think this are important are:

* Ideally it would be nice to fix the problem and land a PR again to  fix https://github.com/certbot/certbot/issues/6901.
* None of our tests caught the problem before the release. It'd be nice to know more details about what went wrong so we can try and prevent similar issues in the future.

**Comment by bmw** [comment]
Due to the recommendation from people who work on Apache, I consolidated what we know so far in an email the Apache developer's mailing list. I've included the relevant bits below so we have all the information in one place too:

Last week, we did a release of Certbot which had it configure httpd to include a configuration file setting “SSLSessionTickets off” in each HTTPS virtual host managed by Certbot. Unfortunately, the addition of this directive caused errors during the TLS handshake in some clients (including Chrome and Firefox) for some of our users. Some of the users affected by this problem were running:

* httpd 2.4.18 (from Ubuntu 16.04)
* httpd 2.4.25 (from Debian 9)
* httpd 2.4.39 (from Amazon Linux 2)

They were presumably using the version of OpenSSL available in those distributions as well although I haven’t been able to verify that yet.

The affected clients and their error messages were:

* Chrome: ERR_SSL_PROTOCOL_ERROR
* Firefox: SSL_ERROR_RX_UNEXPECTED_NEW_SESSION_TICKET
* Wget: OpenSSL: error:1408E0F4:SSL routines:ssl3_get_message:unexpected message

Clients that were reported to be unaffected were:

* OpenSSL s_client
* Safari

Curl was affected for some users, but not others. One report of a failure with curl provided the output below from which I removed their domain and IP:
```
$ curl -v https://example.com
* Rebuilt URL to: https://example.com/
*   Trying 1.2.3.4...
* TCP_NODELAY set 
* Connected to example.com (1.2.3.4) port 443 (#0)
* schannel: SSL/TLS connection with example.com port 443 (step 1/3)
* schannel: checking server certificate revocation
* schannel: sending initial handshake data: sending 172 bytes...
* schannel: sent initial handshake data: sent 172 bytes
* schannel: SSL/TLS connection with example.com port 443 (step 2/3)
* schannel: failed to receive handshake, need more data
* schannel: SSL/TLS connection with example.com port 443 (step 2/3)
* schannel: encrypted data got 2960
* schannel: encrypted data buffer: offset 2960 length 4096
* schannel: sending next handshake data: sending 126 bytes...
* schannel: SSL/TLS connection with example.com port 443 (step 2/3)
* schannel: encrypted data got 258 
* schannel: encrypted data buffer: offset 258 length 4096
* schannel: next InitializeSecurityContext failed: SEC_E_ILLEGAL_MESSAGE (0x80090326) - This error usually occurs when a fatal SSL/TLS alert is received (e.g. handshake failed). More detail may be available in the Windows System event log.
* Closing connection 0
* schannel: shutting down SSL/TLS connection with example.com port 443 
* schannel: clear security context handle
curl: (35) schannel: next InitializeSecurityContext failed: SEC_E_ILLEGAL_MESSAGE (0x80090326) - This error usually occurs when a fatal SSL/TLS alert is received (e.g. handshake failed). More detail may be available in the Windows System event log.
```

**Comment by cpu** [comment]
For anyone like me that's curious to follow the discussion you can find the mailing list thread here: https://lists.apache.org/thread.html/06b1bda49d1adfdb711be64ba4e2347413712fbd3bac61795fe5ae66@%3Cdev.httpd.apache.org%3E

**Comment by bmw** [comment]
Thanks for sharing that link @cpu. I should have done so before.

To write up some more facts here for myself and others, there is very little information about this on Google. I searched for Firefox's error message (which is the most specific) and "apache" or "httpd". There are only a couple pages of results for both and I went through them all without much benefit. The couple things I did find were:

* This problem may be intermittent as suggested by https://mail-archives.apache.org/mod_mbox/httpd-users/201011.mbox/%3C941437.52886.qm@web114518.mail.gq1.yahoo.com%3E.\
* While I think this is unlikely, it could be related to (OpenSSL calls in?) PHP. https://www.apachelounge.com/viewtopic.php?t=4180 has multiple Apache + PHP users reporting a seemingly unrelated issue about PHP crashing when run under Apache and trying to use OpenSSL, but multiple people in that thread also reported the same Firefox error message. Also, the two reporters of the problem in Certbot on Let's Encrypt's community forum whose domain I was able to connect to both are using phpmyadmin. I set up phpmyadmin on Debian 9 but still could not reproduce the problem 😦

**Comment by bmw** [comment]
My thought for next steps here is to try and get more information from people who reported this problem on the community forum. This issue is spread across multiple threads which are:

* https://community.letsencrypt.org/t/certificate-renewed-but-site-not-visible/99421
* https://community.letsencrypt.org/t/schannel-failed-to-receive-handshake-need-more-data/99400
* https://community.letsencrypt.org/t/ssl-error-after-cert-renew/99430

I don't really want to bury the resolution at the bottom of those threads though so I was planning on creating a new thread and mentioning everyone who was reported to have the issue.

While I have not carefully prepared a list of information I'd like to have, some ideas of things to ask for are:

* OpenSSL version (bonus points if you can get this from Apache rather than libssl which technically might not be the same)
* All mod_ssl directives in the Apache config
* "ssl_error_log" output as suggested at https://lists.apache.org/thread.html/aff542914f82f5e38c463c4b412989f5f1d28da82981cf39a337335f@%3Cdev.httpd.apache.org%3E although I'm not immediately familiar with that.
* A high level description of their setup (e.g. are they using PHP/phpmyadmin?)
* Apache modules loaded
* And the most aggressive at all, all of `/etc/apache2` or `/etc/httpd` (which directory is used depends on the system). This could contain a lot of sensitive information though. Ideas on how to help mitigate this if we want to ask for this regardless is to ask them to email it to us, strip out sensitive information themselves (without stripping out relevant information...), or quickly and hackily dust off https://github.com/certbot/certbot/tree/master/letshelp-certbot to help them with this step. For the latter, the main thing that'd have to be done assuming it still works is to not delete the tarball when the process exits and give them instructions on how to install the package.

**Comment by ohemorange** [comment]
Other things that do not account for the issue (at least individually):
- using a CDN
- having one virtual host set to on and another to use
- using apache (we have one report of nginx now as well)
- using AWS

**Comment by ohemorange** [comment]
Also, IE is [reportedly unaffected](https://community.letsencrypt.org/t/errors-from-browsers-with-ssl-session-tickets-off-nginx/18124/5).

**Comment by notroj** [comment]
The point made here https://community.letsencrypt.org/t/ssl-error-after-cert-renew/99430/6 is interesting.  I'm unfamiliar with how certbot creates httpd config - when it was disabling `SSLSessionTickets` was it done so at global context in httpd, or inside a `VirtualHost` block?

**Comment by notroj** [comment]
Also worth noting that at least some of failures linked are showing curl built with SChannel as the SSL toolkit, plausibly this is related to interactions between particular SSL toolkit client side (or versions of).  "curl --version" output is definitely useful to capture since it shows the SSL library used.

**Comment by ohemorange** [comment]
> The point made here https://community.letsencrypt.org/t/ssl-error-after-cert-renew/99430/6 is interesting. I'm unfamiliar with how certbot creates httpd config - when it was disabling SSLSessionTickets was it done so at global context in httpd, or inside a VirtualHost block?

Inside a VirtualHost block.

**Comment by ohemorange** [comment]
Update: current hypothesis is that this bug is due to using old versions of OpenSSL, along with a particular config setup.

Testing with Nginx, we saw the following behavior:
1.0.1t -- broken
1.0.2g -- broken
1.0.2s -- working
1.1.0k -- working
1.1.1c -- working

Note that `openssl version` doesn't necessarily give the version that Nginx is built against, but `nginx -V` will.

The way to reproduce is to have two server blocks. The first one should have `ssl_session_tickets` set to `on` (including implicitly by default, in Nginx), and the second (in the order they're processed) set to `off`.

Particularly due to the timing, we suspect this issue fixed it: https://git.openssl.org/?p=openssl.git;a=commitdiff;h=5c753de
Meaning that this is fixed with versions of openssl > 1.0.2h.

Next step is to test on Apache, by finding a version of Apache built against OpenSSL <= 1.0.2h, but with Apache version >= 2.4.11. Ubuntu xenial should work: Apache version: 2.4.18, openssl version: 1.0.2g. When testing this, it's important to check the version Apache is using by running `grep OpenSSL /var/log/apache2/error.log` rather than `openssl version`, because they might not be the same.

**Comment by notroj** [comment]
Nice job tracking that down!

**Comment by april** [comment]
Do we know when it was fixed in the 1.1.x branch at all?

**Comment by april** [comment]
I've updated the [Mozilla SSL Configuration Generator](https://ssl-config.mozilla.org) to no longer recommend `SSLSessionTickets off` unless OpenSSL is 1.0.2l or higher. Thanks for everyone's hard work on this!

**Comment by ohemorange** [comment]
Currently working on putting out 0.37.2 that stops turning off session tix in Nginx, like we did for Apache.

As for next steps, I'm inclined to match @april's fix for the config generator -- gate on OpenSSL version.

Potential complications for that include:
- Parsing openssl version from `nginx -V`. Printing openssl version info was added in Nginx 1.8.0, so that should be fine, as long as we remember to check for the "running with" [parenthetical](https://github.com/nginx/nginx/blob/master/src/core/nginx.c#L435).
- Parsing openssl version from apache (`grep OpenSSL /var/log/apache2/error.log` or `grep OpenSSL /var/log/apache2/error.log` depending on the system). We'd need to confirm there's something in there first. Shouldn't be too bad.

There's also the option of Do Something Fancy about the systems with the bug -- even with the bug, we *could* still make it work if we turned session tickets off in *all* virtual hosts, or checked about ordering, but honestly not sure it's worth the finicky code for systems running older software that will be updated eventually. Like, if you're using OpenSSL < 1.0.2l, is forward secrecy really your biggest problem? I'd be surprised.

**Comment by bmw** [comment]
I'm closing this issue since @ohemorange was able to figure out the cause of these errors.

I created #7350 and #7351 based on her suggestions above to track disabling session tickets again in Apache and Nginx.

This has been heavily implied above but to say it explicitly for anyone who stumbles upon this thread in the future, the bug described in https://github.com/certbot/certbot/issues/7322#issuecomment-523244023 was fixed in OpenSSL 1.0.2l and above. The fix here was included in the initial release of OpenSSL 1.1.0. You can see this by looking at OpenSSL's git history and seeing that the commit https://github.com/openssl/openssl/commit/5c753de668322bf9903a49ba713b2cbc62667571 is in the history for the `OpenSSL_1_1_0 tag` and https://github.com/openssl/openssl/commit/dbdb96617cce2bd4356d57f53ecc327d0e31f2ad is in the history for the `OpenSSL_1_0_2l` and no prior 1.0.2 tags.

To say echo this one more time, great work @ohemorange!
